designs/2023-test-rule-errors/README.md

```diff
@@ -0,0 +1,309 @@
+
+For example, with the rule [no-restricted-imports](https://eslint.org/docs/rules/no-restricted-imports), one might want to test that the rule schema fails validation when passed:
+
+- Something that isn't an array
+- An empty array
+- An array containing an item that isn't a string nor object
+- An array containing an object that is missing required properties like `name`, or has unknown properties
+- Any other invalid combinations of input
+
+Note that the goal is not to test that [ajv](https://github.com/ajv-validator/ajv) (which performs the validation using JSON Schema) itself works properly, as we can treat that third-party dependency as a blackbox, but instead to test that the rule author has written out their schema correctly.
+
+It can be tricky to get schemas to perfectly represent what the allowed input should be, especially when no automated testing exists around this today. Schemas are often too lenient, allowing extra input that the rule implementation ignores or doesn't handle properly. This can result in rules silently misbehaving, not respecting the wishes or intentions of the user's configuration, or just crashing.
+
+By enabling testing of schemas, we'll make it easier to write higher-quality schemas and rules, thus improving the user experience.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+Add support to the rule tester for a new `fatal` test case array in addition to `valid` and `invalid`.
+
+Example with a schema validation error because a non-existent option is passed:
+
+```json
+{
+    "fatal": [
+        {
+            "code": "foo();",
+            "options": [{ "checkFoo": true, "nonExistentOption": true }],
+            "error": {
+                "message": "Value {\"checkFoo\":true,\"nonExistentOption\":true} should NOT have additional properties.",
+                "name": "SchemaValidationError"
+            }
+        }
+    ]
+}
+```
+
+To test a custom exception thrown by a rule, simply use the exception text as the message: `"message": "Some custom exception message from the rule."` and/or the name of the exception class as the `name`.
+
+The `error` object in the test case uses the `name` and `message` properties directly from the original exception object, while grouping them nicely together in the same structure as the original exception.  Similar to `message` and `messageId` in `invalid` test cases, only one of `message` and `name` is required. If the user doesn't care to test the full exception `message` text, the exception class `name` can be used instead.
```

Comment:
Can you clarify what exactly the `name` property asserts? "The exception class `name`" sounds like we'll check `error.constructor.name`, but the implementation PR checks `err.name`.

Reply:
Good point. Clarified that it should use `err.constructor.name` which is always available based on the actual classname.

Reply:
Actually, that might be the wrong choice. `name` is a standard property that exception classes can set, without having to create a new class.

```js
const e = new Error("Malformed input"); // e.name is 'Error'

e.name = "ParseError";
throw e;
```

Note that if you do create a custom class, you have to fill in this property yourself:

```js
class SchemaValidationError extends Error {
    constructor(message) {
        super(message);
        this.name = "SchemaValidationError";
    }
}
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/name

Which do you think we should use?

Reply:
Good point. Clarified that we should use:

> `name` comes from `err.name` or `err.constructor.name` in case either has a custom value (besides just `Error`)

```js
const name = err.name === 'Error' ? err.constructor.name : err.name;
```

That way, the user can set `name` manually, or create a custom class, and either way we'll use any custom name available.

```js
const e = new Error("Malformed input"); // e.name is 'Error'

e.name = "ParseError"; // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/name
throw e;
```

```js
class SchemaValidationError extends Error {}
```